land_grab_2/stl_dataset/step_2_5/get_cessions.py

```diff
@@ -0,0 +1,278 @@
+
+    Returns:
+    GeoDataFrame -- STL parcels with each parcel-cession intersection as a dis-
+                    tinct row
+    """
+    # Reproject cessions_gdf to match parcels_gdf.
+    cessions_gdf = cessions_gdf.to_crs(parcels_gdf.crs)
+
+    # Subset cessions_gdf to only include the CESSNUM and geometry columns.
+    cessions_gdf = cessions_gdf[["CESSNUM", "geometry"]]
+
+    out_gdf = parcels_gdf.sjoin(cessions_gdf, how="left")
+    out_gdf.drop(columns=["index_right"], inplace=True)
+
+    return out_gdf
+
+
+def aggregate_cessions_by_parcel(
+    parcels_gdf: gpd.GeoDataFrame, cession_codebook_df: pd.DataFrame
+) -> gpd.GeoDataFrame:
+    """
+    Aggregate cession information for each parcel.
+
+    This function aggregates all parcel-cession intersections into a single row,
+    one per parcel. Cession numbers are space-separated in the all_cession_numb-
+    ers field. Additionally, each parcel will have 8 cession number fields,
+    8 present day tribe fields, and 8 historical tribe fields.
+
+    Arguments:
+    parcels_gdf -- GeoDataFrame containing parcel geometries
+    cession_codebook_df -- DataFrame containing cession information
+
+    Returns:
+    GeoDataFrame -- STL parcels with cession information aggregated by parcel
+    """
+
+    def expand_cession_numbers(group):
+        # Split the all_cession_numbers field into a list.
+        cession_list = group["all_cession_numbers"].split()
+
+        # Create fields for tracking individual cession numbers, present day
+        # tribes, and tribes named in cessions.
+        cession_fields = {}
+        for i in range(8):
+            cession_num = f"cession_num_{i+1:02d}"
+            present_day_tribe = f"C{i+1:02d}_present_day_tribe"
+            historical_tribe = f"C{i+1:02d}_tribe_named_in_land_cessions_1784-1894"
+
+            cession_fields[cession_num] = (
+                cession_list[i] if i < len(cession_list) else None
+            )
+
+            # Look up the present day and historical tribes for each cession
+            # number, if available.
+            if cession_fields[cession_num] is not None:
+                matching_cession = cession_codebook_df.loc[
+                    cession_codebook_df["Cession_Number"] == cession_fields[cession_num]
+                ]
+
+                if not matching_cession.empty:
+                    cession_fields[present_day_tribe] = matching_cession.iloc[0][
+                        "Present_Day_Tribe"
+                    ]
+                    cession_fields[historical_tribe] = matching_cession.iloc[0][
+                        "Tribe_Named_in_Land_Cessions_1784-1894"
+                    ]
+                else:
+                    cession_fields[present_day_tribe] = None
+                    cession_fields[historical_tribe] = None
+            else:
+                cession_fields[present_day_tribe] = None
+                cession_fields[historical_tribe] = None
```

Comment:
This is where we do the generation of fields like `cession_num_0{n}`, `C0{n}_present_day_tribe`, `C0{n}_tribe_named_in_land_cessions_1784-1894` for n [1, 8].